### DISCUSSION TITLE
ESBMC does not detect race conditions with floating points

### DISCUSSION THREAD

**Discussion by SanjanaSai94** [esbmc/esbmc#1657] [Q&A]
I run  ESBMC version 7.4.0 64-bit x86_64 windows and it was not able to detect race conditions for the attached code. also I added some assert(0) statements (line 215 and 228) in order to check that is possible to access the shared variable and it also cannot detect these assert(0). the code is a little simple, only two threads and does not have loops. 
Please can you check.
Many thanks for the support.
[main_float.txt](https://github.com/esbmc/esbmc/files/14224502/main_float.txt)

**Comment by SanjanaSai94** [comment]
It also does not work for int numbers

[main_int.txt](https://github.com/esbmc/esbmc/files/14224658/main_int.txt)

**Comment by lucasccordeiro** [comment]
Thanks for this report, @SanjanaSai94.

@XLiZHI: could you please check this issue?

**Comment by XLiZHI** [comment]
Hi @SanjanaSai94 ,

Thanks for this issue.

After my detailed investigation, the key point should be here:

```
    int speed_input = nondet_int();

    if (speed_input > 50) {

        printf("Speed limit exceeded\n");

        return 1;

    }
```
It returns when `speed_input` is greater than 50, this causes subsequent code to be skipped.

`!condition` always be false, and `speed_out` will not be written.

Maybe you could try commenting `return 1;`, like this

```
    int speed_input = nondet_int();

    if (speed_input > 50) {

        printf("Speed limit exceeded\n");

        // return 1;

    }
```
`esbmc main.c --data-races-check --context-bound 2
`
```
----------------------------------------------------
Violated property:
  file main3.c line 216 column 9 function esc_control
  W/W data race on c:@speed_output
  !tmp_c:@speed_output


VERIFICATION FAILED
```

**Comment by lucasccordeiro** [comment]
@XLiZHI: It looks like ESBMC is facing some issues with guards for the Reachability Tree (RT) exploration. We should be able to detect this data-race because nondet_int can also return a `seed_input <= 50`, which would trigger this data-race.

@Anthonysdu: could you please check this issue? I'm almost sure that we've some issues with the guard generation for the RT exploration.

**Reply by XLiZHI** [reply]
Hi @lucasccordeiro,

But when seed_imput < 50, the assignment OUTPUT won't happen.

I think ESBMC is working as expected.

**Reply by lucasccordeiro** [reply]
Oh, I see. Then, I agree with your assessment here.
